rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java

```diff
@@ -106,120 +130,68 @@ public NativeArrayBuffer slice(double s, double e) {
     }
 
-    // Function-calling dispatcher
-
-    @Override
-    public Object execIdCall(
-            IdFunctionObject f, Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
-        if (!f.hasTag(CLASS_NAME)) {
-            return super.execIdCall(f, cx, scope, thisObj, args);
-        }
-        int id = f.methodId();
-        switch (id) {
-            case ConstructorId_isView:
-                return Boolean.valueOf(
-                        (isArg(args, 0) && (args[0] instanceof NativeArrayBufferView)));
-
-            case Id_constructor:
-                double length = isArg(args, 0) ? ScriptRuntime.toNumber(args[0]) : 0;
-                return new NativeArrayBuffer(length);
-
-            case Id_slice:
-                NativeArrayBuffer self = realThis(thisObj, f);
-                double start = isArg(args, 0) ? ScriptRuntime.toNumber(args[0]) : 0;
-                double end = isArg(args, 1) ? ScriptRuntime.toNumber(args[1]) : self.buffer.length;
-                return self.slice(start, end);
-        }
-        throw new IllegalArgumentException(String.valueOf(id));
+    private static NativeArrayBuffer getSelf(Scriptable thisObj) {
+        return LambdaConstructor.convertThisObject(thisObj, NativeArrayBuffer.class);
     }
 
-    private static NativeArrayBuffer realThis(Scriptable thisObj, IdFunctionObject f) {
-        return ensureType(thisObj, NativeArrayBuffer.class, f);
+    private static NativeArrayBuffer js_constructor(Context cx, Scriptable scope, Object[] args) {
+        double length = isArg(args, 0) ? ScriptRuntime.toNumber(args[0]) : 0;
+        return new NativeArrayBuffer(length);
     }
 
-    private static boolean isArg(Object[] args, int i) {
-        return ((args.length > i) && !Undefined.instance.equals(args[i]));
+    private static Boolean js_isView(
+            Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        return Boolean.valueOf((isArg(args, 0) && (args[0] instanceof NativeArrayBufferView)));
     }
 
-    @Override
-    protected void initPrototypeId(int id) {
-        String s;
-        int arity;
-        switch (id) {
-            case Id_constructor:
-                arity = 1;
-                s = "constructor";
-                break;
-            case Id_slice:
-                arity = 2;
-                s = "slice";
-                break;
-            default:
-                throw new IllegalArgumentException(String.valueOf(id));
+    private static NativeArrayBuffer js_slice(
+            Context cx,
+            Scriptable scope,
+            Scriptable thisObj,
+            LambdaConstructor defaultConstructor,
+            Object[] args) {
+        NativeArrayBuffer self = getSelf(thisObj);
+        double start = isArg(args, 0) ? ScriptRuntime.toNumber(args[0]) : 0;
+        double end = isArg(args, 1) ? ScriptRuntime.toNumber(args[1]) : self.getLength();
+        int endI =
+                ScriptRuntime.toInt32(
```

Comment:
I may be wrong here, but aren't most of the values here -- "self.getLength()" -- integers already? And is this method supposed to work if "start" and "end" are not integers? It looks like we are converting a bunch of integers into a floating-point value and then back to integers again.

Or are the "start" and "end" parameters here actually doubles because they could theoretically represent a value larger than 2^31? And if so can we handle that with our native byte arrays anyway?

In other words I'm asking if this whole thing shouldn't be done with integer arithmetic, and there may well be a reason!

Reply:
I'll have to look into it. I think I just copy pasted what was done before, but I'll see.

Reply:
It does allow for non-integers but they are supposed to get converted to integers before all the max/mins in the following lines.

Reply:
However, after seeing what tests failed, you were right. It looks like they can be more than 2^31, but in that case, the value would be greater than the length of the byte array, so when taking the min, it will just be the original byte array length, avoiding any issues.